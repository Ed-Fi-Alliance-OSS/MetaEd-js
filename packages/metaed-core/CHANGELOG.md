# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.1.1-dev.35"></a>
## [1.1.1-dev.35](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.34...v1.1.1-dev.35) (2018-02-14)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.34"></a>
## [1.1.1-dev.34](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.32...v1.1.1-dev.34) (2018-01-18)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.33"></a>
## [1.1.1-dev.33](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.32...v1.1.1-dev.33) (2018-01-18)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.32"></a>
## [1.1.1-dev.32](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.31...v1.1.1-dev.32) (2018-01-17)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.31"></a>
## [1.1.1-dev.31](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.30...v1.1.1-dev.31) (2018-01-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.30"></a>
## [1.1.1-dev.30](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.29...v1.1.1-dev.30) (2018-01-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.29"></a>
## [1.1.1-dev.29](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.26...v1.1.1-dev.29) (2018-01-14)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.28"></a>
## [1.1.1-dev.28](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.26...v1.1.1-dev.28) (2017-12-20)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.27"></a>
## [1.1.1-dev.27](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.26...v1.1.1-dev.27) (2017-12-20)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.26"></a>
## [1.1.1-dev.26](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.25...v1.1.1-dev.26) (2017-12-18)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.25"></a>
## [1.1.1-dev.25](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.23...v1.1.1-dev.25) (2017-12-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.24"></a>
## [1.1.1-dev.24](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.23...v1.1.1-dev.24) (2017-12-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.23"></a>
## [1.1.1-dev.23](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.22...v1.1.1-dev.23) (2017-12-13)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.22"></a>
## [1.1.1-dev.22](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.18...v1.1.1-dev.22) (2017-12-13)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.20"></a>
## [1.1.1-dev.20](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.18...v1.1.1-dev.20) (2017-12-13)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.19"></a>
## [1.1.1-dev.19](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.18...v1.1.1-dev.19) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.18"></a>
## [1.1.1-dev.18](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.16...v1.1.1-dev.18) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.17"></a>
## [1.1.1-dev.17](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.16...v1.1.1-dev.17) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.16"></a>
## [1.1.1-dev.16](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.15...v1.1.1-dev.16) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.15"></a>
## [1.1.1-dev.15](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.14...v1.1.1-dev.15) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.14"></a>
## [1.1.1-dev.14](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.13...v1.1.1-dev.14) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.13"></a>
## [1.1.1-dev.13](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.12...v1.1.1-dev.13) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.12"></a>
## [1.1.1-dev.12](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.11...v1.1.1-dev.12) (2017-12-10)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.11"></a>
## [1.1.1-dev.11](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.10...v1.1.1-dev.11) (2017-12-09)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.10"></a>
## [1.1.1-dev.10](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.9...v1.1.1-dev.10) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.9"></a>
## [1.1.1-dev.9](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.8...v1.1.1-dev.9) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.8"></a>
## [1.1.1-dev.8](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.7...v1.1.1-dev.8) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.7"></a>
## [1.1.1-dev.7](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.6...v1.1.1-dev.7) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.6"></a>
## [1.1.1-dev.6](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.5...v1.1.1-dev.6) (2017-12-04)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.5"></a>
## [1.1.1-dev.5](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.4...v1.1.1-dev.5) (2017-11-29)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.4"></a>
## [1.1.1-dev.4](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.3...v1.1.1-dev.4) (2017-11-27)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.2"></a>
## [1.1.1-dev.2](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.1...v1.1.1-dev.2) (2017-11-27)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.1"></a>
## [1.1.1-dev.1](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.0...v1.1.1-dev.1) (2017-11-27)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.0"></a>
## [1.1.1-dev.0](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.0.0-dev.5...v1.1.1-dev.0) (2017-11-21)




**Note:** Version bump only for package metaed-core
