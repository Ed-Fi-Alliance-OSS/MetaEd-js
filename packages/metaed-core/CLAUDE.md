# CLAUDE.md

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## MetaEd Core Package Overview

This is the core package of the MetaEd system, providing the compilation pipeline, entity models, and plugin infrastructure for transforming MetaEd DSL files into various output formats.

## Essential Commands

**Testing:**
- `npm test` - Run all tests for this package
- `npx jest path/to/test.ts` - Run specific test file
- `npm run test:coverage` - Run with coverage report

**Development:**
- `npm run build` - Build the package
- `npm run antlr-gen` - Regenerate ANTLR parser from grammar

## Architecture Overview

### Pipeline Flow

The `Pipeline.ts` orchestrates these sequential stages:
1. **Initialize** - Set up environment and plugins
2. **Load Files** - Read .metaed files from filesystem
3. **Validate Syntax** - Parse with ANTLR grammar
4. **Build Parse Tree** - Create ANTLR parse trees
5. **Walk Builders** - Convert parse trees to entity models
6. **Run Plugins** - Execute validators → enhancers → generators
7. **Write Output** - Save generated artifacts

### Core Components

**State Management:**
- `State.ts` - Central pipeline state containing configuration, files, entities, and results
- `MetaEdEnvironment.ts` - Runtime environment with namespaces, entities, and plugin data

**Entity System:**
- `model/` - Entity definitions (Association, DomainEntity, Common, etc.)
- `model/property/` - Property types (String, Integer, Boolean, Association, etc.)
- All entities extend `ModelBase` → `TopLevelEntity`

**Builder Pattern:**
- `builder/` - ANTLR listeners that construct entities from parse trees
- One builder per entity type (e.g., `DomainEntityBuilder`, `AssociationBuilder`)
- Two-phase processing: enter/exit methods for tree walking

**Plugin Architecture:**
- **Validators**: `(metaEd: MetaEdEnvironment) => ValidationFailure[]`
- **Enhancers**: `(metaEd: MetaEdEnvironment) => EnhancerResult`
- **Generators**: `(metaEd: MetaEdEnvironment) => Promise<GeneratorResult>`
- Plugins run in dependency order
- Validators and enhancers in each plugin run in an explicit order provided by the plugin `initialize()` function 

### Key Patterns

1. **Entity Repository**: Typed maps within namespaces for entity storage/retrieval
2. **Immutable Updates**: Never modify existing entities, create new versions
3. **Plugin Data Isolation**: Each plugin stores data under `data[pluginName]`
4. **Validation Failures**: Include source location for error reporting

### Testing Approach

Use `MetaEdTextBuilder` to construct test MetaEd code:
```typescript
MetaEdTextBuilder.build()
  .withBeginNamespace('TestNamespace')
  .withStartDomainEntity('TestEntity')
  .withStringProperty('TestProperty')
  .withEndDomainEntity()
  .sendToListener(builder);
```

### Important Files

- `index.ts` - Main exports (extensive public API)
- `Pipeline.ts` - Pipeline orchestration
- `MetaEdGrammar.g4` - ANTLR grammar definition
- `WalkBuilders.ts` - Builder orchestration logic
- `EntityRepository.ts` - Entity storage/retrieval

### Development Notes

- Grammar changes require running `npm run antlr-gen`
- New entity types need: model class, builder, and tests
- Validators should return descriptive error messages with source locations
- Enhancers must only add data, never mutate existing properties.
- Generators are async and return file content/metadata