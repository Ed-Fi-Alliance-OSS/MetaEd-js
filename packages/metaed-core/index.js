// @flow

// Test helper -- *** where should this go? ***
export { MetaEdTextBuilder } from './test/MetaEdTextBuilder';


// Functions

export { deepFreeze, deepFreezeAssign, prependIndefiniteArticle } from './src/Utility';
export { startingFromFileLoadP } from './src/task/Pipeline';
export { createMetaEdFile } from './src/task/MetaEdFile';
export { loadCoreBufferedFiles, loadExtensionBufferedFiles } from './src/task/BufferFileLoader';
export { loadFiles } from './src/task/FileSystemFilenameLoader';
export { newState } from './src/State';
export { addEntity, getAllEntities, getAllTopLevelEntities, getEntity, getTopLevelCoreEntity, getEntitiesOfType, getAllEntitiesNoSimpleTypes } from './src/model/EntityRepository';
export { addProperty, getAllProperties, getPropertiesOfType, newPropertyIndex } from './src/model/property/PropertyRepository';
export { isReferenceProperty, isSharedProperty, allPropertyTypes } from './src/model/property/PropertyType';
export { allEntityModelTypes, allTopLevelEntityModelTypes, topLevelCoreEntityModelTypes } from './src/model/ModelType';

// Builders

export { AssociationBuilder } from './src/builder/AssociationBuilder';
export { AssociationExtensionBuilder } from './src/builder/AssociationExtensionBuilder';
export { AssociationSubclassBuilder } from './src/builder/AssociationSubclassBuilder';
export { ChoiceBuilder } from './src/builder/ChoiceBuilder';
export { CommonBuilder } from './src/builder/CommonBuilder';
export { CommonExtensionBuilder } from './src/builder/CommonExtensionBuilder';
export { DecimalTypeBuilder } from './src/builder/DecimalTypeBuilder';
export { DescriptorBuilder } from './src/builder/DescriptorBuilder';
export { DomainBuilder } from './src/builder/DomainBuilder';
export { DomainEntityBuilder } from './src/builder/DomainEntityBuilder';
export { DomainEntityExtensionBuilder } from './src/builder/DomainEntityExtensionBuilder';
export { DomainEntitySubclassBuilder } from './src/builder/DomainEntitySubclassBuilder';
export { EnumerationBuilder } from './src/builder/EnumerationBuilder';
export { IntegerTypeBuilder } from './src/builder/IntegerTypeBuilder';
export { InterchangeBuilder } from './src/builder/InterchangeBuilder';
export { NamespaceInfoBuilder } from './src/builder/NamespaceInfoBuilder';
export { SharedDecimalBuilder } from './src/builder/SharedDecimalBuilder';
export { SharedIntegerBuilder } from './src/builder/SharedIntegerBuilder';
export { SharedStringBuilder } from './src/builder/SharedStringBuilder';
export { StringTypeBuilder } from './src/builder/StringTypeBuilder';

// Flow types

export type { State } from './src/State';
export type { InputDirectory } from './src/task/FileSystemFilenameLoader';
export type { FileSet, MetaEdFile } from './src/task/MetaEdFile';
export type { FileIndex } from './src/task/FileIndex';
export type { Validator } from './src/validator/Validator';
export type { ValidationFailure } from './src/validator/ValidationFailure';
export type { Enhancer } from './src/enhancer/Enhancer';
export type { EnhancerResult } from './src/enhancer/EnhancerResult';
export type { Generator } from './src/generator/Generator';
export type { GeneratorResult } from './src/generator/GeneratorResult';
export type { GeneratedOutput } from './src/generator/GeneratedOutput';
export type { MetaEdEnvironment, PluginEnvironment, SemVer } from './src/MetaEdEnvironment';
export type { MetaEdPlugin, PluginManifest } from './src/plugin/PluginTypes';

// Properties
export type { AssociationProperty } from './src/model/property/AssociationProperty';
export type { BooleanProperty } from './src/model/property/BooleanProperty';
export type { ChoiceProperty } from './src/model/property/ChoiceProperty';
export type { CommonProperty } from './src/model/property/CommonProperty';
export type { CurrencyProperty } from './src/model/property/CurrencyProperty';
export type { DateProperty } from './src/model/property/DateProperty';
export type { DecimalProperty } from './src/model/property/DecimalProperty';
export type { DescriptorProperty } from './src/model/property/DescriptorProperty';
export type { DomainEntityProperty } from './src/model/property/DomainEntityProperty';
export type { DurationProperty } from './src/model/property/DurationProperty';
export type { EntityProperty, HasReferencedEntity } from './src/model/property/EntityProperty';
export type { EnumerationProperty } from './src/model/property/EnumerationProperty';
export type { InlineCommonProperty } from './src/model/property/InlineCommonProperty';
export type { IntegerProperty } from './src/model/property/IntegerProperty';
export type { MergedProperty } from './src/model/property/MergedProperty';
export type { PercentProperty } from './src/model/property/PercentProperty';
export type { PropertyIndex } from './src/model/property/PropertyRepository';
export type { PropertyType } from './src/model/property/PropertyType';
export type { ReferentialProperty } from './src/model/property/ReferentialProperty';
export type { SchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export type { SharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export type { SharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export type { SharedShortProperty } from './src/model/property/SharedShortProperty';
export type { SharedStringProperty } from './src/model/property/SharedStringProperty';
export type { ShortProperty } from './src/model/property/ShortProperty';
export type { SimpleProperty } from './src/model/property/SimpleProperty';
export type { SharedSimple } from './src/model/SharedSimple';
export type { StringProperty } from './src/model/property/StringProperty';
export type { TimeProperty } from './src/model/property/TimeProperty';
export type { YearProperty } from './src/model/property/YearProperty';

// Entities
export type { AbstractEntity } from './src/model/AbstractEntity';
export type { Association } from './src/model/Association';
export type { AssociationExtension } from './src/model/AssociationExtension';
export type { AssociationSubclass } from './src/model/AssociationSubclass';
export type { Choice } from './src/model/Choice';
export type { Common } from './src/model/Common';
export type { CommonExtension } from './src/model/CommonExtension';
export type { DecimalType } from './src/model/DecimalType';
export type { Descriptor } from './src/model/Descriptor';
export type { Domain } from './src/model/Domain';
export type { DomainEntity } from './src/model/DomainEntity';
export type { DomainEntityExtension } from './src/model/DomainEntityExtension';
export type { DomainEntitySubclass } from './src/model/DomainEntitySubclass';
export type { DomainItem } from './src/model/DomainItem';
export type { EntityRepository } from './src/model/EntityRepository';
export type { Enumeration } from './src/model/Enumeration';
export type { EnumerationItem } from './src/model/EnumerationItem';
export type { IntegerType } from './src/model/IntegerType';
export type { Interchange } from './src/model/Interchange';
export type { InterchangeExtension } from './src/model/InterchangeExtension';
export type { InterchangeItem } from './src/model/InterchangeItem';
export type { MapTypeEnumeration } from './src/model/MapTypeEnumeration';
export type { ModelBase } from './src/model/ModelBase';
export type { ModelType } from './src/model/ModelType';
export type { NamespaceInfo } from './src/model/NamespaceInfo';
export type { SchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export type { SharedDecimal } from './src/model/SharedDecimal';
export type { SharedInteger } from './src/model/SharedInteger';
export type { SharedString } from './src/model/SharedString';
export type { SourceMap } from './src/model/SourceMap';
export type { StringType } from './src/model/StringType';
export type { Subdomain } from './src/model/Subdomain';
export type { TopLevelEntity } from './src/model/TopLevelEntity';

// Null Objects
export { NoMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { NoTopLevelEntity } from './src/model/TopLevelEntity';

// Factories
export { newMetaEdEnvironment, newPluginEnvironment } from './src/MetaEdEnvironment';
export { newMetaEdPlugin } from './src/plugin/PluginTypes';

export { newAssociationProperty } from './src/model/property/AssociationProperty';
export { newBooleanProperty } from './src/model/property/BooleanProperty';
export { newChoiceProperty } from './src/model/property/ChoiceProperty';
export { newCommonProperty } from './src/model/property/CommonProperty';
export { newCurrencyProperty } from './src/model/property/CurrencyProperty';
export { newDateProperty } from './src/model/property/DateProperty';
export { newDecimalProperty } from './src/model/property/DecimalProperty';
export { newDescriptorProperty } from './src/model/property/DescriptorProperty';
export { newDomainEntityProperty } from './src/model/property/DomainEntityProperty';
export { newDurationProperty } from './src/model/property/DurationProperty';
export { newEnumerationProperty } from './src/model/property/EnumerationProperty';
export { newInlineCommonProperty } from './src/model/property/InlineCommonProperty';
export { newIntegerProperty } from './src/model/property/IntegerProperty';
export { newMergedProperty } from './src/model/property/MergedProperty';
export { newPercentProperty } from './src/model/property/PercentProperty';
export { newSchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export { newSharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export { newSharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export { newSharedShortProperty } from './src/model/property/SharedShortProperty';
export { newSharedStringProperty } from './src/model/property/SharedStringProperty';
export { newShortProperty } from './src/model/property/ShortProperty';
export { newSimpleProperty } from './src/model/property/SimpleProperty';
export { newStringProperty } from './src/model/property/StringProperty';
export { newTimeProperty } from './src/model/property/TimeProperty';
export { newYearProperty } from './src/model/property/YearProperty';

export { newAbstractEntity } from './src/model/AbstractEntity';
export { newAssociation } from './src/model/Association';
export { newAssociationExtension } from './src/model/AssociationExtension';
export { newAssociationSubclass } from './src/model/AssociationSubclass';
export { newChoice } from './src/model/Choice';
export { newCommon } from './src/model/Common';
export { newCommonExtension } from './src/model/CommonExtension';
export { newDecimalType } from './src/model/DecimalType';
export { newDescriptor } from './src/model/Descriptor';
export { newDomain } from './src/model/Domain';
export { newDomainEntity } from './src/model/DomainEntity';
export { newDomainEntityExtension } from './src/model/DomainEntityExtension';
export { newDomainEntitySubclass } from './src/model/DomainEntitySubclass';
export { newDomainItem } from './src/model/DomainItem';
export { newEnumeration } from './src/model/Enumeration';
export { newEnumerationItem } from './src/model/EnumerationItem';
export { newInlineCommon } from './src/model/Common';
export { newIntegerType } from './src/model/IntegerType';
export { newInterchange } from './src/model/Interchange';
export { newInterchangeExtension } from './src/model/InterchangeExtension';
export { newInterchangeItem } from './src/model/InterchangeItem';
export { newMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { newNamespaceInfo } from './src/model/NamespaceInfo';
export { newSchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export { newSharedDecimal } from './src/model/SharedDecimal';
export { newSharedInteger } from './src/model/SharedInteger';
export { newSharedString } from './src/model/SharedString';
export { newSourceMap } from './src/model/SourceMap';
export { newStringType } from './src/model/StringType';
export { newSubdomain } from './src/model/Subdomain';
export { asTopLevelEntity, newTopLevelEntity } from './src/model/TopLevelEntity';
