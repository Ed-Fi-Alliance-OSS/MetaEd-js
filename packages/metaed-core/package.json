{
  "name": "metaed-core",
  "main": "dist/index.js",
  "version": "2.1.3-dev.3",
  "description": "MetaEd core functionality plugin",
  "license": "SEE LICENSE IN LICENSE.md",
  "publishConfig": {
    "registry": "https://www.myget.org/F/ed-fi/npm/"
  },
  "files": [
    "/dist",
    "/CHANGELOG.md",
    "/LICENSE.md",
    "/package.json"
  ],
  "dependencies": {
    "ajv": "^6.0.0",
    "antlr4": "4.6.1-patched.0",
    "chalk": "^2.3.0",
    "cosmiconfig": "^5.0.5",
    "deep-freeze": "^0.0.1",
    "deepmerge": "^2.1.1",
    "final-fs": "^1.6.1",
    "fs-extra": "^5.0.0",
    "joi": "12.0.0",
    "klaw-sync": "^3.0.2",
    "ramda": "^0.25.0",
    "semver": "^5.4.1",
    "sugar": "^2.0.4",
    "topo": "2.0.2",
    "winston": "3.0.0"
  },
  "devDependencies": {
    "normalize-path": "^2.1.1"
  },
  "scripts": {
    "build": "yarn build:clean && yarn build:copy-non-ts && yarn build:dist",
    "build:clean": "rimraf dist",
    "build:dist": "tsc",
    "build:copy-non-ts": "copyfiles -u 1 -e \"**/*.ts\" \"src/**/*\" dist --verbose"
  }
}
