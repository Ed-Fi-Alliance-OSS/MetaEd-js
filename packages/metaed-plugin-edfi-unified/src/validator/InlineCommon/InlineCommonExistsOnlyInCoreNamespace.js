// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.common.forEach(common => {
    if (!common.inlineInOds || !common.namespaceInfo.isExtension) return;
    failures.push({
      validatorName: 'InlineCommonExistsOnlyInCoreNamespace',
      category: 'error',
      message: `${common.typeHumanizedName} ${common.metaEdName} is not valid in extension namespace ${common.namespaceInfo.projectExtension} .`,
      sourceMap: common.sourceMap.type,
      fileMap: null,
    });
  });
  return failures;
}
