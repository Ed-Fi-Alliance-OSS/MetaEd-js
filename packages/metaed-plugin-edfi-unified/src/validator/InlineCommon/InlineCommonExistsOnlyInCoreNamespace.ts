// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace) => {
    if (!namespace.isExtension) return;
    namespace.entity.common.forEach((common) => {
      if (!common.inlineInOds || !common.namespace.isExtension) return;
      failures.push({
        validatorName: 'InlineCommonExistsOnlyInCoreNamespace',
        category: 'error',
        message: `${common.typeHumanizedName} ${common.metaEdName} is not valid in extension namespace ${common.namespace.namespaceName}.`,
        sourceMap: common.sourceMap.metaEdName,
        fileMap: null,
      });
    });
  });
  return failures;
}
