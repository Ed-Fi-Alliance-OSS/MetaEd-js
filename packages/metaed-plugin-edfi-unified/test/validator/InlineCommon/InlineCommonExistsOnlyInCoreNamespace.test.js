// @flow
import { newMetaEdEnvironment, MetaEdTextBuilder, CommonBuilder, NamespaceInfoBuilder } from 'metaed-core';
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { validate } from '../../../src/validator/InlineCommon/InlineCommonExistsOnlyInCoreNamespace';

describe('when validating inline common type exists in core', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartInlineCommon('EntityName')
      .withDocumentation('EntityDocumentation')
      .withBooleanProperty('PropertyName', 'PropertyDocumentation', true, false)
      .withEndInlineCommon()
      .withEndNamespace()

      .sendToListener(new NamespaceInfoBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one common', () => {
    expect(metaEd.entity.common.size).toBe(1);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating inline common type exists in extension', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('extension', 'ProjectExtension')
      .withStartInlineCommon('EntityName')
      .withDocumentation('EntityDocumentation')
      .withBooleanProperty('PropertyName', 'PropertyDocumentation', true, false)
      .withEndInlineCommon()
      .withEndNamespace()

      .sendToListener(new NamespaceInfoBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one common', () => {
    expect(metaEd.entity.common.size).toBe(1);
  });

  it('should have validation failure', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('InlineCommonExistsOnlyInCoreNamespace');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot(
      'when validating inline common type exists in extension should have validation failure -> message',
    );
    expect(failures[0].sourceMap).toMatchSnapshot(
      'when validating inline common type exists in extension should have validation failure -> sourceMap',
    );
  });
});
