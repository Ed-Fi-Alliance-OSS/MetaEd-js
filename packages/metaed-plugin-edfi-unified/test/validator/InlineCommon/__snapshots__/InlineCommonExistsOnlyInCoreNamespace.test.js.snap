// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating inline common type exists in extension should have validation failure 1`] = `"Inline Common EntityName is not valid in extension namespace ProjectExtension ."`;

exports[`when validating inline common type exists in extension should have validation failure 2`] = `
Object {
  "column": 2,
  "line": 12,
  "tokenText": "Inline Common",
}
`;
